config_v2.md

```diff
@@ -0,0 +1,225 @@
+                       config=configuration)
+                       
+def my_main_function(batch_size, dataset_size, nb_epochs, run, logger):
+    # main experiment here
+    ...
+
+with ex.start():
+    my_main_function(**ex.config, run=ex.current_run, logger=ex.logger)
+```
+
+
+### Example with delayed evaluation of a config variable:
+
+
+This handles the case where a configuration value, if not set should be evaluated based on other values present in the config.
+
+If the `Delayed` object is present even after all overrides from the command lines and other, it is evaluated before starting the Experiment.
+
+```python
+import sacred
+from sacred import Delayed
+
+
+configuration = dict(
+    batch_size=32, 
+    dataset_size=Delayed(lambda config: config['batch_size'] * 100), 
+    nb_epochs=50
+)
+
+
+ex = sacred.Experiment('my_pretty_experiment',
+                       config=configuration)
+                       
+def my_main_function(batch_size, dataset_size, nb_epochs):
+    # main experiment here
+    ...
+
+with ex.start():
+    my_main_function(**ex.config)
+```
+
+For example:
+
+```bash
+python my_main.py with batch_size=60  # dataset_size is 6000 here
+python my_main.py with dataset_size=5000  # dataset_size is 5000 here
+```
+
+
+##### Api of the Delayed object:
+
+```python
+class Delayed:
+    def __init__(self, evaluation_function, priority=0):
+        """
+        The evaluation function takes the config dict as an input and 
+        returns an object which will go in the config.
+        
+        In cases where some Delayed object must be evaluated in a certain order, 
+        because they depend on some other Delayed objects,
+        we can use the priority argument to make sure that the Delayed objects are
+        evaluated correctly, in order. If the priority of two delayed objects are 
+        equal, we use the depth or the order of the list to choose which goes first.
+        Starting from py3.7, we'll use the order of the dict too.
+        
+        With python 3.7, dict are ordered by default and we will use
+        this to avoid specifying 'priority' in some other cases. priority will
+        only be used in some very weird (but certainly valid) cases.
+        """
+        ...
+
+```
+
+### Example with selecting part of the configuration from the command line
+
+This is a replacement for named configs.
+
+
+```python
+import sacred
+
+
+configuration = dict(batch_size=32, dataset_size=10_000, nb_epochs=50)
+
+def config_change1(config):
+    config['dataset_config'] = dict(crop_size=(30, 30), random_flip=True)
+    config['dataset_size'] = 500
+    
+
+def config_change2(config):
+    config['dataset_config'] = dict(crop_size=(50, 50), random_flip=False)
+    config['dataset_size'] = 700
+
+
+ex = sacred.Experiment('my_pretty_experiment',
+                       config=configuration,
+                       potential_modifications=[config_change1, config_change2])
```

Comment:
Having to declare all the modifiers feels a bit clunky; I quite liked the old decorator approach. It's a bit magical but I think worth it for convenience.

If we are going to go for "zero-magic", should we declare the name of the modifications as well (rather than presumably getting it via inspection)?

Reply:
> Having to declare all the modifiers feels a bit clunky; I quite liked the old decorator approach. It's a bit magical but I think worth it for convenience.

Very good point. I believe I have another solution than the decorators which will save us from having an enormous list of potential modifications. No magic but no big list.  I'll update the proposal to adress your point. 

> If we are going to go for "zero-magic", should we declare the name of the modifications as well (rather than presumably getting it via inspection)?

Also a very good point. We should have the possibility to use 0 magic. I believe it should be optional. When not specified, we would use `__name__`. The question is whether we use a decorative to specify the name, or we specify the name when passing it to the config/experiment constructor. To have a consistent API (see #572 and #569), I'll believe it's better that we use an optional decorator to specify the name of this config modification.